[export] Allow constant attr mutation

angelayi opened the pull request:
Test Plan: CI

Differential Revision: D56893728


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

avikchaudhuri commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -2498,7 +2498,11 @@ def inline_call_(
+                (
+                    variables.CustomizedDictVariable,
+                    variables.UserDefinedObjectVariable,
+                    variables.NNModuleVariable,
```
What does this do? It's hard to wrap my head around the condition, but is it that you can now define `__setattr__` on a `nn.module`? 

If so does it get inlined? 

Why was it banned before?

angelayi replied:
Ah sorry, I think this was part of another direction I was trying to take to fix the const attr issue -- I deleted this change and looks like the tests still work 😅

avikchaudhuri commented on test/dynamo/test_export_mutations.py:
```diff
@@ -57,18 +44,6 @@ def forward(self, x):
 
         self.check_failure_on_export(Foo(), torch.randn(3, 2))
 
-    def test_module_attribute_mutation_violation_positive_4(self):
```
So this works now? Just gets inlined into a constant on the RHS, is that why? 

But I don't see `self.a` being initialized, so isn't that a new attribute, like `test_module_attribute_mutation_violation_positive_3`? Confused about the mental model here.

angelayi replied:
Yeah, self.a is a new attribute, but it's a constant like in `mutation_violation_positive_3`. `mutation_violation_positive_3` is failing due to a different error, so I moved it to `test_export` instead.

angelayi committed:
[export] Allow constant attr mutation (#125424)

Summary:

This PR enables attribute mutations for only constant attributes. The behavior matches non-strict, where the constants will be directly embedded in the graph, and the original module will be mutated after exporting.

Test Plan: CI

Reviewed By: pianpwk

Differential Revision: D56893728
